File: backend/secure_research/audit.py
from __future__ import annotations

import json
from pathlib import Path
from threading import Lock
from typing import Any


class JsonlAuditLog:
    def __init__(
        self,
        directory: Path,
        *,
        max_bytes: int = 5 * 1024 * 1024,
        backups: int = 3,
    ) -> None:
        self.directory = directory
        self.path = directory / "research-audit.jsonl"
        self.max_bytes = max_bytes
        self.backups = backups
        self._lock = Lock()

    def append(self, record: dict[str, Any]) -> None:
        self.directory.mkdir(parents=True, exist_ok=True)
        payload = json.dumps(record, separators=(",", ":"), sort_keys=True)
        with self._lock:
            self._rotate_if_needed(len(payload) + 1)
            with self.path.open("a", encoding="utf-8") as handle:
                handle.write(payload)
                handle.write("\n")

    def _rotate_if_needed(self, incoming_bytes: int) -> None:
        if not self.path.exists():
            return
        if self.path.stat().st_size + incoming_bytes <= self.max_bytes:
            return

        oldest = self.path.with_suffix(f".jsonl.{self.backups}")
        if oldest.exists():
            oldest.unlink()
        for index in range(self.backups - 1, 0, -1):
            source = self.path.with_suffix(f".jsonl.{index}")
            if source.exists():
                source.replace(self.path.with_suffix(f".jsonl.{index + 1}"))
        self.path.replace(self.path.with_suffix(".jsonl.1"))

File: backend/client/src/runs/audit.ts
import { appendFile, mkdir, rename, stat, unlink } from "node:fs/promises";
import { join } from "node:path";
import type { TranscriptEvent } from "@secure-research/contracts";

export class MetadataAudit {
  private readonly path: string;

  constructor(
    directory: string,
    private readonly maxBytes = 5 * 1024 * 1024,
    private readonly backups = 3,
  ) {
    this.path = join(directory, "research-audit.jsonl");
    void mkdir(directory, { recursive: true });
  }

  async append(event: TranscriptEvent): Promise<void> {
    const record = {
      timestamp: event.timestamp,
      turnId: event.turnId,
      sequence: event.sequence,
      channel: event.channel,
      origin: event.origin,
      direction: event.direction,
      kind: event.kind,
      method: event.method,
      requestId: event.requestId,
      parentRequestId: event.parentRequestId,
      level: event.level,
      status: event.status,
      summary: event.summary,
      durationMs: event.durationMs,
      metadata: event.metadata,
    };
    const line = `${JSON.stringify(record)}\n`;
    await this.rotateIfNeeded(Buffer.byteLength(line));
    await appendFile(this.path, line, { encoding: "utf8" });
  }

  private async rotateIfNeeded(incomingBytes: number): Promise<void> {
    try {
      const current = await stat(this.path);
      if (current.size + incomingBytes <= this.maxBytes) return;
    } catch {
      return;
    }

    const oldest = `${this.path}.${this.backups}`;
    await unlink(oldest).catch(() => undefined);
    for (let index = this.backups - 1; index >= 1; index -= 1) {
      await rename(`${this.path}.${index}`, `${this.path}.${index + 1}`).catch(
        () => undefined,
      );
    }
    await rename(this.path, `${this.path}.1`);
  }
}
